kep/real-load-aware-scheduling/real-load-aware-scheduling.md

```diff
@@ -0,0 +1,538 @@
+
+
+### Goals
+
+1. Minimize the number of running nodes.
+2. CPU Utilisation per node should not go beyond X%.
+3. Instances of an app should spread across failure domains. The failure domain would be host at the lowest level and can extend to racks, zones, regions, etc.
+4. To not affect the behavior of default score plugins unless necessary.
+
+### Non-Goals
+
+1. The constraints above are the best effort of soft constraints. They are not hard constraints.
+2. Descheduling or pre-emption due to lousy scheduling decisions is not addressed in the initial design.
+3. Memory, Network, and Disk utilization are not considered in the initial design.
+
+## Proposal
+
+
+### User Stories
+
+
+#### Story 1
+
+
+As a company relying on the public cloud, we would like to minimize machine costs by efficiently utilizing the nodes leased.
+
+
+#### Story 2
+
+As a company-owned data center, we would like to minimize the hardware and maintenance associated with clusters by getting the maximum possible juice out of existing machines.
+
+
+#### Story 3
+
+As a K8s user, I would like to see more enriched single-level scheduler options in the community.
+
+
+### Notes/Constraints/Caveats
+
+Enabling our plugin(s) will cause conflict with 2 default scoring plugins: “NodeResourcesLeastAllocated” & “NodeResourcesBalancedAllocation” score plugins. So it is strongly advised to disable them.
+
+
+### Risks and Mitigations
+
+If utilization metrics are not available for a long time, we will fall back to the best fit bin pack based on allocations. There is no user action needed for this.
+
+
+## Design Details
+
+
+
+<img src="images/design.png" alt="design" width="720" height="628"/>
+
+
+
+### Metrics Provider
+
+This service provides metrics backed by a time-series database — for example, Prometheus, InfluxDB, etc.
+
+
+### Load Watcher/Load Analyser
+
+The load watcher and analyzer both run in a single process. The watcher is responsible for the cluster-wide aggregation of resource usage metrics like CPU, memory, network, and IO stats over windows of a specified duration. It stores these in its local cache and persists several aggregations in the host local DB for fault tolerance. The analyzer is responsible for the detection of bad metrics and any remediation. Bad metrics could be due to missing metrics or metrics with considerable errors, making them anomalies. It can also be extended in the future to use ML models for analyzing metrics.
+
+Load watcher would cache metrics in the last 15-minute, 10-minute, and 5-minute windows, which can be queried via REST API exposed.
+
+
+### DB
+
+This is a localhost database stored as a file. This DB aims to populate the load watcher's cache quickly if it is lost due to a crash and perform scheduling decisions without any impact on latency.
```

Comment:
If the scheduler is running as a pod or static pod，how to ensure persistence of file when the process crashes.
I think it's better add more detail about this.

Reply:
The file will be stored in host file system, so it will be persisted across pod failures on the same node. For HA setups, each load watcher process will maintain its corresponding file locally. I will add this to doc :+1: